### ISSUE TITLE
In-memory zip generation

### ISSUE THREAD

**Issue by mikeabdullah** [pixelglow/zipzap#8]
In my app, I'd like to generate the data for a zip file entirely in-memory without touching the disk. The data is to be uploaded to a remote server, so there's no _need_ for it to touch the local disk, and avoiding so means no file access errors can occur.

Looking through `ZZMutableArchive`, this looks a bit tricky to achieve. It seems the current code is built atop `NSFileHandle`. I could create a custom subclass of `NSFileHandle` that is backed by data, but that seems rather messy, and perhaps would fail in some places.

Using `NSOutputStream` instead would make my life easier. So, what in `NSFileHandle` are we dependent on? Looking at the code:
- `-offsetInFile` to track how much data has been written so far; this could be replaced by passing round integers as arguments/return values
- If an error occurs while writing an entry, winds the file handle back to where it was before that entry was written
- `-truncateFileAtOffset` called once writing is finished; as I understand it, that mostly is to handle archives which are now smaller than when they were read

Have I got those right, missed anything?

**Comment by RLovelett** [comment]
I agree with @ndusart some example of in-memory reading and writing would be great.

**Comment by ilyakaz** [comment]
I can't find ZZMutableArchive in the current code. I tried initializing a ZZArchive instead in Swift based on the sample code above:

`let zipArchive = try ZZArchive.init(data: NSMutableData())`

It throws this error:

**The operation couldn’t be completed. (com.pixelglow.ZipZap error 0.)
Error Domain=com.pixelglow.ZipZap Code=0 "(null)" UserInfo={NSUnderlyingError=0x14a3a1af0 {Error Domain=NSCocoaErrorDomain Code=260 "The file couldn’t be opened because it doesn’t exist."}}**

**Comment by pixelglow** [comment]
I'll take a look at the error. Meanwhile please raise this as a separate issue.

**Comment by ilyakaz** [comment]
Done. Thank you.


https://github.com/pixelglow/ZipZap/issues/145
